src/v2/Components/__stories__/Molecules.story.tsx

```diff
@@ -476,3 +483,629 @@ export const Content = () => {
     </Page>
   )
 }
+
+const SwiperWithProgress: React.FC<
+  SwiperProps & { verticalAlign?: "top" | "bottom" }
+> = ({ children, verticalAlign = "bottom", ...rest }) => {
+  const count = React.Children.count(children)
+  const [boxProps, swiperProps] = splitBoxProps(rest)
+  const [index, setIndex] = useState(0)
+  const progress = (index * 100) / (count - 1)
+
+  return (
+    <Box {...boxProps}>
+      <FullBleed>
+        <Swiper
+          onChange={setIndex}
+          mb={6}
+          Rail={props => {
+            return (
+              <SwiperRail
+                {...props}
+                alignItems={verticalAlign === "top" ? "flex-start" : "flex-end"}
+                px={4}
+                mx={4}
+              />
+            )
+          }}
+          {...swiperProps}
+        >
+          {React.Children.map(children, (child: any, i) => {
+            return React.cloneElement(child, {
+              ml: i === 0 ? [2, 4] : undefined,
+              mr: i === count - 1 ? [2, 4] : undefined,
+            })
+          })}
+        </Swiper>
+      </FullBleed>
+
+      <CarouselBar percentComplete={progress} />
+    </Box>
+  )
+}
```

Comment:
![](https://static.damonzucconi.com/_capture/0EuvSsrA.gif)

Since this relies on the index it's a bit chunky (though this can be good if we want to turn on snapping). Though one benefit is that we don't have a scroll listener on all of these. We may want to take that approach. Either way is fine.